File: go/021.go
func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
    if l1 == nil {
        return l2
    }
    if l2 == nil {
        return l1
    }
    dummy := &ListNode{}
    tail := dummy
    for l1 != nil && l2 != nil {
        if l1.Val <= l2.Val {
            temp := l1.Next
            l1.Next = nil
            tail.Next = l1
            l1 = temp
        } else {
            temp := l2.Next
            l2.Next = nil
            tail.Next = l2
            l2 = temp
        }
        tail = tail.Next
    }
    if l1 != nil {
        tail.Next = l1
    }
    if l2 != nil {
        tail.Next = l2
    }
    return dummy.Next
}

File: cpp/021.cpp
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* l1, ListNode* l2) {
        if (l1 == nullptr)
            return l2;
        if (l2 == nullptr)
            return l1;
        ListNode *dummy = new ListNode(0);
        ListNode *tail = dummy;
        while (l1 != nullptr && l2 != nullptr) {
            if (l1->val <= l2->val) {
                ListNode *temp = l1->next;
                l1->next = nullptr;
                tail->next = l1;
                l1 = temp;
            } else {
                ListNode *temp = l2->next;
                l2->next = nullptr;
                tail->next = l2;
                l2 = temp;
            }
            tail = tail->next;
        }
        if (l1 != nullptr)
            tail->next = l1;
        if (l2 != nullptr)
            tail->next = l2;
        return dummy->next;
    }
};
